contracts/peg/NonCustodialPSM.sol

```diff
@@ -0,0 +1,456 @@
+/// When funds are needed for a redemption, they are simply pulled from the PCV Deposit
+contract NonCustodialPSM is
+    INonCustodialPSM,
+    RateLimited,
+    OracleRef,
+    ReentrancyGuard
+{
+    using Decimal for Decimal.D256;
+    using SafeCast for *;
+    using SafeERC20 for IERC20;
+
+    /// @notice the fee in basis points for selling asset into FEI
+    uint256 public override mintFeeBasisPoints;
+
+    /// @notice the fee in basis points for buying the asset for FEI
+    uint256 public override redeemFeeBasisPoints;
+
+    /// @notice the PCV deposit target to deposit and withdraw from
+    IPCVDeposit public override pcvDeposit;
+
+    /// @notice the token this PSM will exchange for FEI
+    /// This token will be set to WETH9 if the bonding curve accepts eth
+    IERC20 public immutable override underlyingToken;
+
+    /// @notice Rate Limited Minter contract that will be called when FEI needs to be minted
+    GlobalRateLimitedMinter public immutable rateLimitedMinter;
+
+    /// @notice the max mint and redeem fee in basis points
+    /// Governance can change this fee
+    uint256 public override MAX_FEE = 300;
+
+    /// @notice boolean switch that indicates whether redemptions are paused
+    bool public redeemPaused;
+
+    /// @notice boolean switch that indicates whether minting is paused
+    bool public mintPaused;
+
+    /// @notice event that is emitted when redemptions are paused
+    event RedemptionsPaused(address account);
+
+    /// @notice event that is emitted when redemptions are unpaused
+    event RedemptionsUnpaused(address account);
+
+    /// @notice event that is emitted when minting is paused
+    event MintingPaused(address account);
+
+    /// @notice event that is emitted when minting is unpaused
+    event MintingUnpaused(address account);
+
+    /// @notice struct for passing constructor parameters related to OracleRef
+    struct OracleParams {
+        address coreAddress;
+        address oracleAddress;
+        address backupOracle;
+        int256 decimalsNormalizer;
+        bool doInvert;
+    }
+
+    /// @notice struct for passing constructor parameters related to MultiRateLimited
+    struct RateLimitedParams {
+        uint256 maxRateLimitPerSecond;
+        uint256 rateLimitPerSecond;
+        uint256 bufferCap;
+    }
+
+    struct PSMParams {
+        uint256 mintFeeBasisPoints;
+        uint256 redeemFeeBasisPoints;
+        uint256 feiLimitPerSecond;
+        uint256 mintingBufferCap;
+        IERC20 underlyingToken;
+        IPCVDeposit pcvDeposit;
+        GlobalRateLimitedMinter rateLimitedMinter;
+        uint112 feiRateLimitPerSecond;
+        uint144 feiBufferCap;
+        uint112 underlyingTokenRateLimitPerSecond;
+        uint144 underlyingTokenBufferCap;
+    }
+
+    /// @notice constructor
+    /// @param params PSM constructor parameter struct
+    constructor(
+        OracleParams memory params,
```

Comment:
Using calldata here should save a lot of gas

Reply:
can't make these constructor params calldata, compiler yells when you try